openssl/src/crypto/symm_new.rs

```diff
@@ -0,0 +1,1835 @@
+            Error::Unspecified => "Unspecified openssl error",
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        if let Error::IoError(ref err) = *self {
+            write!(fmt, "{}: {}", self.description(), err)
+        }
+        else {
+            write!(fmt, "{}", self.description())
+        }
+    }
+}
+
+impl error::FromError<IoError> for Error {
+    fn from_error(err: IoError) -> Error {
+        Error::IoError(Box::new(err))
+    }
+}
+
+macro_rules! chk {
+    ($inp:expr) => (
+        {
+            let r = $inp;
+            assert!(r == 1);
+            r
+        }
+    );
+}
+
+/// A common cipher interface
+struct Context {
+    ctx: *mut ffi::EVP_CIPHER_CTX,
+    state: State,
+}
+
+/// A trait for ciphers that allow transcoding several chunks of data consequtively.
+pub trait Apply {
+    /// Transcode the `data` into the `buf`.
+    ///
+    /// The `buf` have enough space to fit `data` (plus a cipher block length
+    /// in ECB and CBC modes).
+    fn apply(&mut self, data: &[u8], buf: &mut [u8]) -> usize;
+}
+
+/// A trait for block mode ciphers that may not return the last bytes of data until finished.
+pub trait PaddedFinish: Apply {
+    fn finish(&mut self, buf: &mut [u8]) -> Result<usize, Error>;
+}
+
+/*
+/// A cipher that works on large blocks (sectors)
+trait SectorMode {
+    fn apply(&mut self, iv: &[u8], data: &[u8], buf: &mut [u8]);
+}
+*/
+
+/// An adapter for using ciphers as `Writer`s
+// Subject to changes after std::io stabilization
+pub struct Filter<'a, T: 'a> {
+    cipher: &'a mut T,
+    sink: &'a mut (Writer + 'a),
+}
+
+const FILTER_BUFFER_LEN: usize = 16384;
+
+impl <'a, T> Filter<'a, T> {
+    /// Create a `Writer` adapter for the `cipher`. The output is written
+    /// to the `sink`.
+    /// The `cipher` has to be `start`ed beforehand and `finish`ed after
+    /// destroying the adapter.
+    pub fn new(cipher: &'a mut T, sink: &'a mut (Writer + 'a))
+          -> Filter<'a, T> {
+        Filter { cipher: cipher, sink: sink }
+    }
+}
+
+impl <'a, T: Apply> Writer for Filter<'a, T> {
+    fn write_all(&mut self, data: &[u8]) -> Result<(), IoError> {
+        let mut buf = [0; FILTER_BUFFER_LEN + ffi::EVP_MAX_BLOCK_LENGTH];
+        for chunk in data.chunks(FILTER_BUFFER_LEN) {
+            let len = self.cipher.apply(chunk, &mut buf);
+            if len > 0 {
+                try!(self.sink.write_all(&buf[..len]));
+            }
+        }
+        Ok(())
+    }
+}
+
+/// A `Writer` adapter that finishes the padded cipher after writing.
+pub struct PaddedFilter<'a, T: 'a> {
+    inner: Filter<'a, T>,
+    closed: bool,
+}
+
+impl <'a, T: PaddedFinish> PaddedFilter<'a, T> {
+    /// Create a `Writer` adapter that finishes the padded cipher after writing.
+    /// The output is written to the `sink`.
+    /// The cipher has to be `start`ed beforehand and is finished explicitly
```

Comment:
This was inspired by Rust RFC discussions of having explicit `close` that can return errors. I'm not aware of a final decision there. We need to propagate a malformed padding error somewhere.

So the choice here is between having a `close() -> Result<(), Error>` on the filter and checking for errors in its parent struct with `finish() -> Result<(), Error>` or `get_last_error() -> Option<Error>`. But the `PaddedFilter` can't leave the cipher unfinished because its sink is going away. Checking with `finish` after the filter is dropped would be the only (and extra confusing) case of calling `finish` on a finished cipher -- just to check for an error. `get_last_error` speaks for itself.